drivers/ble/sphero/sphero_r2d2_driver.go

```diff
@@ -0,0 +1,474 @@
+// TODO get this working for R2
+func (d *R2D2Driver) SetDataStreamingConfig(dsc spherocommon.DataStreamingConfig) error {
+	buf := new(bytes.Buffer)
+	if err := binary.Write(buf, binary.BigEndian, dsc); err != nil {
+		return err
+	}
+	d.sendCraftPacket(buf.Bytes(), 0x02, 0x11)
+	return nil
+}
+
+// initialize tells driver to get ready to do work
+func (d *R2D2Driver) initialize() error {
+	if err := d.antiDOSOff(); err != nil {
+		return err
+	}
+	if err := d.Wake(); err != nil {
+		return err
+	}
+	// subscribe to Sphero response notifications
+	if err := d.Adaptor().Subscribe(r2ResponseChara, d.handleResponses); err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			packet := <-d.packetChannel
+			err := d.writeCommand(packet)
+			if err != nil {
+				d.Publish(d.Event(spherocommon.ErrorEvent), err)
+			}
+		}
+	}()
+
+	d.ConfigureCollisionDetection(d.defaultCollisionConfig)
+	// enableStopOnDisconnect is a temporary option, not supported by the R2
+	// d.enableStopOnDisconnect()
+
+	return nil
+}
+
+// antiDOSOff turns off Anti-DOS code so we can control R2D2
+func (d *R2D2Driver) antiDOSOff() error {
+	str := "usetheforce...band"
+	buf := &bytes.Buffer{}
+	buf.WriteString(str)
+
+	if err := d.Adaptor().WriteCharacteristic(r2AntiDosChara, buf.Bytes()); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (d *R2D2Driver) writeCommand(packet *packet) error {
+	log.Printf("request %X % X % X %X %X\n", sop, packet.header, packet.body, packet.checksum, eop)
+
+	buf := append([]uint8{sop}, escapeBytes(packet.header)...)
+	buf = append(buf, escapeBytes(packet.body)...)
+	buf = append(buf, escapeByte(packet.checksum)...)
+	buf = append(buf, eop)
+
+	if err := d.Adaptor().WriteCharacteristic(r2CommandsChara, buf); err != nil {
+		fmt.Println("async send command error:", err)
+		return err
+	}
+
+	// avoid ddos the r2
+	time.Sleep(commandInterval)
+
+	return nil
+}
+
+// shutdown stops R2D2 driver (void)
+func (d *R2D2Driver) shutdown() error {
+	d.Sleep()
+	time.Sleep(750 * time.Microsecond)
+	return nil
+}
+
+// handleResponses handles responses returned from R2D2
+func (d *R2D2Driver) handleResponses(data []byte) {
+	//log.Printf("handleResponse of %v bytes: % X", len(data), data)
+
+	// v2 packets can be arbitrary length, we have to puzzle them together
+	// they also are sent in 1 byte chunks and can be out of order
+
+	// ignore end of packet and wait till we see another packet
+	if len(data) > 0 && data[0] == eop {
+		return
+	}
+
+	// append message parts to existing
+	if len(data) > 0 && data[0] != sop {
+		d.asyncBuffer = append(d.asyncBuffer, data...)
+		return
+	}
+
+	// clear message when new one begins (first byte is always 0x8D)
+	if len(data) > 0 {
+		// append end of packet to complete message
+		d.asyncMessage = append(d.asyncBuffer, eop)
+		d.asyncBuffer = data
+	}
+
+	log.Printf("processing asyncMessage: % X", d.asyncMessage)
+
+	//TODO get sensor data from a packet starting with 0x8d 0x0 0x18 0x2 0xff
```

Comment:
more work needs to be done here. I could see us doing something similar to the python lib were the response stores local sensor variables that can be used by clients calling a simple GetLocation(): X, Y that returns the stored sensor data. A sample sensor packet looks like this:
```
0x8d 0x0 0x18 0x2 0xff 0x41 0x8a 0x5e 0x2f 0x3e 0x61 0x39 0xe3 0x3f 0x2c 0xf5 0x95 0x3b 0xcf 0x4 0x60 0xbe 0x9b 0x13 0xc0 0x3f 0x78 0x68 0x65 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x80 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0xbd 0xfa 0xe7 0xa4 0xbd 0xa2 0xee 0xda 0x0 0x0 0x0 0x0 0xd7 0xd8
```